### PR TITLE
CDC/etcd_worker: add rate limiter to limit EtcdWorker tick frequency

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve? 
#3112 
Too frequent etcd worker tick will cause etcd to be overburdened, and the current etcd qps will increase exponentially with the number of tables that need to be replicated.

### What is changed and how it works?
Add rate limiter to limit EtcdWorker ticks frequency.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

**Summary:** Limiting the tick frequency of `EtcdWoker` can reduce etcd qps by about 50%, but it will reduce the replication speed by about 20%.

**Test Environment:**  
1. 2 Machines: 8-core 16G memory 500G high-speed disk
2. Sink Type: mysql 
3. Cluster topo: 
```yaml
pd_servers:
  - host: 172.x.x.x
  - host: 172.x.x.x

tidb_servers:
  - host: 172.x.x.x
  - host: 172.x.x.x
    
tikv_servers:
  - host: 172.x.x.x
  - host: 172.x.x.x

cdc_servers:
  - host: 172.x.x.x
  - host: 172.x.x.x

monitoring_servers:
  - host: 172.x.x.x

grafana_servers:
  - host: 172.x.x.x

alertmanager_servers:
  - host: 172.x.x.x
```

**Test1** (EtcdWorker ticks limit 10 times / s)
Create **16** changefeeds, synchronize a table for each, and write 50w rows of data to each table in upstream.

<!DOCTYPE html><br class="Apple-interchange-newline">

  | Max checkpoint lag | Average checkpoint lag in peak | Max etcd qps | CPU usage | Total synchronization time
-- | -- | -- | -- | -- | --
owner | 3.3 min | 1.1min | 320 | 164% | 4 min
processor | - | - | 160 | 82% | -

**Test2** (EtcdWorker ticks limit 10 times / s)
Create **30** changefeeds, synchronize a table for each, and write 50w rows of data to each table in upstream.

<!DOCTYPE html><br class="Apple-interchange-newline">

  | Max checkpoint lag | Average checkpoint lag in peak | Max etcd qps | CPU usage | Total synchronization time
-- | -- | -- | -- | -- | --
owner | 6.8 min | 5 min | 600 | 189% | 10 min
processor | - | - | 300 | 95% | -

**Test3** (EtcdWorker ticks without limit )
Create **16** changefeeds, synchronize a table for each, and write 50w rows of data to each table in upstream.
<!DOCTYPE html><br class="Apple-interchange-newline">

  | Max checkpoint lag | Average checkpoint lag in peak | Max etcd qps | CPU usage | Total synchronization time
-- | -- | -- | -- | -- | --
owner | 2.74 min | 50 s | 630 | 193% | 3 min
processor | - | - | 260 | 92% | -

**Test4** (EtcdWorker ticks without limit )
Create **30** changefeeds, synchronize a table for each, and write 50w rows of data to each table in upstream.

<!DOCTYPE html><br class="Apple-interchange-newline">

  | Max checkpoint lag | Average checkpoint lag in peak | Max etcd qps | CPU usage | Total synchronization time
-- | -- | -- | -- | -- | --
owner | 6.9 min | 3 min | 1180 | 199% | 8 min
processor | - | - | 600 | 100% | -

**Test5** (Incremental scan、EtcdWorker ticks limit 10 times / s) 
1. Create 30 changefeed and pause.
2. 30 tables are created upstream, and 50w data is written into each table.
3. After waiting for the upstream data to be written, resume the above 30 changefeeds.

|           | Max checkpoint lag | Average checkpoint lag in peak | **Max etcd qps** | CPU usage | Total synchronization time |
| --------- | ------------------ | ------------------------------ | ---------------- | --------- | -------------------------- |
| owner     | 7 min              | 4 min                          | 600              | 400%      | 9 min                      |
| processor | -                  | -                              | 300              | 200%      | -                          |

**Test6** (Incremental scan、EtcdWorker ticks limit 10 times / s) 
1. Create 30 changefeed and pause.
2. 30 tables are created upstream, and 50w data is written into each table.
3. After waiting for the upstream data to be written, resume the above 30 changefeeds.

When resume changefeed, the owner was down instantly.

  | Max checkpoint lag | Average checkpoint lag in peak | Max etcd qps | CPU usage | Total synchronization time
-- | -- | -- | -- | -- | --
owner | 10 min | 7 min | 871 | 400% | 10 min
processor | - | - | 658 | 200% | -

Side effects
 - Has exported variable changed.
 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: pkg/orchestrator/etcd_worker.go [hunk 1]
```diff
@@ -120,7 +121,8 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 		sessionDone = make(chan struct{})
 	}
 	lastReceivedEventTime := time.Now()
-
+	// limit the frequency of EtcdWorker ticks
+	rl := rate.NewLimiter(10, 30)
```
Comment (liuzix) [inline pkg/orchestrator/etcd_worker.go:125]: Can you add some comments on the meaning of the numbers? And ideally they should be configurable.

Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go]: don

**Comment by maxshuang** [comment]
the current etcd qps will increase exponentially with the number of tables that need to be replicated.

----- @asddongmen  could u explain why?

**Comment by maxshuang** [comment]
I still can't get the point of your tests.  "write 50w rows of data to each table in upstream" means a heavy continuous flow in upstream,  so it is reasonable to increase the etcdworker ticks to speed up.  

Am I right?   Or merge some resolvedTs or checkpointTs report will be a better solution? instead of just limit the rate.

**Comment by liuzix** [comment]
> I still can't get the point of your tests. "write 50w rows of data to each table in upstream" means a heavy continuous flow in upstream, so it is reasonable to increase the etcdworker ticks to speed up.
> 
> Am I right? Or merge some resolvedTs or checkpointTs report will be a better solution? instead of just limit the rate.

From what I understand, the point of this PR is to address the problem that the frequency of ticks in one node must be greater than or equal to the frequency of updates by all nodes. This behavior is not optimal from a very low level.

If you mean batching updates for different changefeeds by "merge some resolvedTs or checkpointTs report", I think this will be a different optimization. The behavior of the business logic is orthogonal to what we are trying to fix here. We can investigate this and address it in another PR.

File: pkg/orchestrator/etcd_worker.go [hunk 2]
```diff
@@ -121,6 +122,10 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 	}
 	lastReceivedEventTime := time.Now()
 
+	// tickRate represents the number of times EtcdWorker can tick
+	// the reactor per second
+	tickRate := time.Second / timerInterval
+	rl := rate.NewLimiter(rate.Limit(tickRate), int(tickRate)*3)
```
Comment (liuzix) [inline pkg/orchestrator/etcd_worker.go:128]: Could you explain the choice of `int(tickRate)*3`, and add a comment here?

**Review by overvenus — APPROVED** [review]

File: pkg/orchestrator/etcd_worker.go [hunk 3]
```diff
@@ -184,6 +188,13 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 			if err := worker.applyUpdates(); err != nil {
 				return errors.Trace(err)
 			}
+
+			// if !rl.Allow(), skip this Tick to avoid etcd worker tick reactor too frequency
+			// It make etcdWorker to batch etcd changed event in worker.state
```
Comment (liuzix) [inline pkg/orchestrator/etcd_worker.go:193]:
```
The semantics of `ReactorState` requires that any implementation can batch updates internally.
```
Can you add this in comments?

**Review by liuzix — APPROVED** [review]

**Comment by hi-rustin** [comment]
@asddongmen You don't seem to have filled out the release note correctly, so please fill out the release note for the next fix.

**Pull request merged**